packages/dds/merge-tree/docs/Attribution.md

```diff
@@ -0,0 +1,378 @@
+This optimization would be entirely transparent to the public API: whatever snapshot/in-memory format we use, we'd always convert to `AttributionInfo` before
+returning the information for a given seq to the DDS/application.
+
+Interfaces might look like this, with exported properties being those visible to an application:
+
+```typescript
+export interface AttributionInfo {
+  user: IUser;
+  timestamp: number;
+}
+
+export interface IAttributor {
+  getAttributionInfo(seq: number): AttributionInfo;
+}
+
+type InternedRef = number & { readonly InternedRef: 'e86840d8-8384-450c-b0e3-9a2855ba2d21'};
+
+interface ObjectInterner {
+  getOrCreateRef(obj: Jsonable): InternedRef;
+  getObject(id: InternedRef): Jsonable;
+  getSerializable(): Jsonable;
+}
+
+interface CompactAttributionInfo {
+  userRef: InternedRef;
+  timestamp: number;
+}
+
+// Concrete types for a particular `Attributor` implementation
+interface SerializedAttributor {
+  interner: Jsonable /* result of calling getSerializable() on an ObjectInterner */
+  lookup: {
+    [seq: number]: InternedRef /* to CompactAttributionInfo */ | CompactAttributionInfo
+  }
+}
+```
+
+Another way to accomplish similar gains without the need to compare `IUser`s would be to maintain a historical `clientId -> IUser` lookup and use `clientId` in place of `userRef`.
+This would be less compact but more readable.
+
+#### Adjacent Range Coalescing
+
+Typical documents will likely have a number of consecutive ops with the same attribution information.
+This happens for a few reasons: users might make a number of edits in a short period of time (consider a user typing
+out a new paragraph), and ops submitted by a single container are batched under some circumstances.
+
+Rather than end up with a `SerializedAttributor` that resembles this:
+
+```javascript
+{
+  interner: [{ email: "john.doe@contoso.com", id: "f400ddf3-4d04-48e9-8783-4b1db8a45fc3" }, { user: 0, timestamp: 1661974200000 }],
+  lookup: {
+    50: 1,
+    51: 1,
+    52: 1,
+    53: 1,
+    54: 1
+  }
+}
+```
+
+we could instead serialize the lookup table like so:
+
+```javascript
+{
+  interner: [{ email: "john.doe@contoso.com", id: "f400ddf3-4d04-48e9-8783-4b1db8a45fc3" }, { user: 0, timestamp: 1661974200000 }],
+  lookup: [{ key: [50, 54], value: 1 }]
+}
+```
+
+Since objects are distinguishable from numbers, single-number ranges could just have a number key.
+
+```typescript
+interface AttributionEntry {
+  /**
+   * Either a single `seq` number for this attribution entry, or a consecutive range `[start, end]` (inclusive)
+   * of `seq` numbers which all have the same attribution information.
+   */
+  k: number | [number, number];
+  v: InternedRef | CompactAttributionInfo;
+}
+
+// Concrete types for a particular `Attributor` implementation
+interface SerializedAttributor {
+  interner: Jsonable; /* result of calling getSerializable() on an ObjectInterner */
+  lookup: AttributionEntry[];
+}
+```
+
+#### Timestamp Binning
```

Comment:
i'm not sure any kind of binning is really necessary or saves much. basically seq + timestamp can be stored as two parallel arrays where a seq's index matches the same index for the timestamp. these two arrays can then be very efficiently compressed: https://www.timescale.com/blog/time-series-compression-algorithms-explained/

we can use the data here to play with different ways of storing data to see what we'd want: https://github.com/microsoft/FluidFrameworkTestData

Reply:
you could even add a third array here for user info id's, that will have lots of duplicates and should also compress well. we're basically doing columnar compression if you think about the data as rows with columns: seq, timestamp, userid(seq)

Reply:
maybe rather than binning we should just think about precision, which is similar, but different. this could be configured per dds for how they store timestamps, 1u, 10ms, 1s, 1m .... this would only affect individual timestamps, so the lower the precision the more duplicates, less diff, better compression. the interesting thing is you could change it any time without any changes, as the mapping from seq to timestamp is still 1 to 1. you could even imagine a scheme where we drop precision over time, so older attribution data is less precise, but more compressible

Reply:
I really like the metapoint here of viewing things as time series and explaining in terms of existing compression schemes. That will be especially handy in terms of code size / reusability if it works well with some of Justus's work on op compression.

> maybe rather than binning we should just think about precision, which is similar, but different. this could be configured per dds for at what precision they store timestamps, 1u, 10ms, 1s, 1m .... this would only affect individual timestamps, so the lower the precision the more duplicates, less diff, better compression. the interesting thing is you could change it any time without any other changes, as the mapping from seq to timestamp is still 1 to 1. you could even imagine a scheme where we drop precision over time, so older attribution data is less precise, but more compressible

This description aligns pretty closely with what I was trying to express with the binning section. So there's probably a gap somewhere in the document :). I'm guessing some wording I had was confusing, could you point me to it?

Reply:
i think it was just the binning language, and prominence of the topic. as i think with run length encoding, even high precision (ms) is probably fine data wise. The code sample is also misleading, as it loses date information which i think we will almost always want.